commands/operator-sdk/cmd/scorecard/test_definitions.go

```diff
@@ -0,0 +1,293 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scorecard
+
+import (
+	"context"
+
+	olmapiv1alpha1 "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Type Definitions
+
+type Test interface {
+	GetName() string
+	GetDescription() string
+	IsCumulative() bool
+	Run(context.Context) *TestResult
+}
+
+type TestResult struct {
+	Test          Test
+	EarnedPoints  int
+	MaximumPoints int
+	Suggestions   []string
+	Errors        []error
+}
+
+type TestInfo struct {
+	Name        string
+	Description string
+	// If a test is set to cumulative, the scores of multiple runs of the same test on separate CRs are added together for the total score.
+	// If cumulative is false, if any test failed, the total score is 0/1. Otherwise 1/1.
+	Cumulative bool
+}
+
+// Any struct that embeds TestInfo only needs to
+// implement Run to implement the Test interface
+func (i TestInfo) GetName() string        { return i.Name }
+func (i TestInfo) GetDescription() string { return i.Description }
+func (i TestInfo) IsCumulative() bool     { return i.Cumulative }
+
+type BasicTestConfig struct {
+	Client   client.Client
+	CR       *unstructured.Unstructured
+	ProxyPod *v1.Pod
+}
+
+type OLMTestConfig struct {
+	Client   client.Client
+	CR       *unstructured.Unstructured
+	CSV      *olmapiv1alpha1.ClusterServiceVersion
+	CRDsDir  string
+	ProxyPod *v1.Pod
+}
+
+type TestSuite struct {
+	TestInfo
+	Tests       []Test
+	TestResults []*TestResult
+	Weights     map[string]float64
+}
+
+// Test definitions
+
+type CheckSpecTest struct {
+	TestInfo
+	BasicTestConfig
+}
+
+func NewCheckSpecTest(conf BasicTestConfig) *CheckSpecTest {
+	return &CheckSpecTest{
+		BasicTestConfig: conf,
+		TestInfo: TestInfo{
+			Name:        "Spec Block Exists",
+			Description: "Custom Resource has a Spec Block",
+			Cumulative:  false,
+		},
+	}
+}
+
+type CheckStatusTest struct {
+	TestInfo
+	BasicTestConfig
+}
+
+func NewCheckStatusTest(conf BasicTestConfig) *CheckStatusTest {
+	return &CheckStatusTest{
+		BasicTestConfig: conf,
+		TestInfo: TestInfo{
+			Name:        "Status Block Exists",
+			Description: "Custom Resource has a Status Block",
+			Cumulative:  false,
+		},
+	}
+}
+
+type WritingIntoCRsHasEffectTest struct {
```

Comment:
Have we considered making files for each type with its own run function next to it? 

It is a minor layout thing, but will make it easier to find the actual test later one IMO.

Reply:
That would be a useful change and make it easier to find things. I think that should be done in a follow up PR though.